zson/context.go

```diff
@@ -285,34 +244,47 @@ func (c *Context) AddColumns(r *zng.Record, newCols []zng.Column, vals []zng.Val
 	return zng.NewRecord(typ, zv), nil
 }
 
-// LookupByName returns the Type indicated by the zng type string.  The type string
+// LookupByName returns the Type indicated by the ZSON type string.  The type string
 // may be a simple type like int, double, time, etc or it may be a set
-// or an array, which are recusively composed of other types.  The set and array
-// type definitions are encoded in the same fashion as zeek stores them as type field
-// in a zeek file header.  Each unique compound type object is created once and
-// interned so that pointer comparison can be used to determine type equality.
-
-//XXX package zson should do this now...?
-
+// or an array, which are recusively composed of other types.  Nested sub-types
+// of complex types are each created once and interned so that pointer comparison
+// can be used to determine type equality.
 func (c *Context) LookupByName(zson string) (zng.Type, error) {
-	typ, err := LookupType(c, zson) //XXX
-	if err != nil {
-		if typ := c.LookupTypeDef(zson); typ != nil {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	typ, ok := c.toType[zson]
+	if !ok {
+		if typ := zng.LookupPrimitive(zson); typ != nil {
+			c.toBytes[typ] = zcode.Bytes(zson)
+			c.toType[zson] = typ
 			return typ, nil
 		}
+		var err error
+		// ParseType() we will re-enter the context and create and/or
+		// return an existing type.  Since it's re-entrant we can't
+		// (and don't want to) hold the lock.  There can be a race
+		// here but it doesn't matter because there is only ever one
+		// type that wins the day because of the incremental locking on
+		// each component of a nested type.
+		c.mu.Unlock()
+		typ, err = ParseType(c, zson)
+		if err != nil {
+			return nil, err
+		}
+		c.mu.Lock()
+		// ParseType will ensure the canonical zson is in the toType table,
+		// but the zson argument here may be any conforming zson type string.
+		// Since this string may appear repeatedly (e.g., type values
+		// coming from an external system) we put an extra entry in the
+		// lookup-table to cache it so we don't parse every instance
+		// of a type string when it is not in canonical form.
+		c.toType[zson] = typ
 	}
-	return typ, err
+	return typ, nil
```

Comment:
Nit: I think this reads better with less nesting.
```go
	if typ, ok := c.toType[zson]; ok {
		return typ, nil
	}
	if typ := zng.LookupPrimitive(zson); typ != nil {
		c.toBytes[typ] = zcode.Bytes(zson)
		c.toType[zson] = typ
		return typ, nil
	}
	// ParseType we will re-enter the context and create and/or
	// return an existing type.  Since it's re-entrant we can't
	// (and don't want to) hold the lock.  There can be a race
	// here but it doesn't matter because there is only ever one
	// type that wins the day because of the incremental locking on
	// each component of a nested type.
	c.mu.Unlock()
	typ, err := ParseType(c, zson)
	if err != nil {
		return nil, err
	}
	c.mu.Lock()
	// ParseType will ensure the canonical zson is in the toType table,
	// but the zson argument here may be any conforming zson type string.
	// Since this string may appear repeatedly (e.g., type values
	// coming from an external system) we put an extra entry in the
	// lookup-table to cache it so we don't parse every instance
	// of a type string when it is not in canonical form.
	c.toType[zson] = typ
	return typ, nil
```